File: backend-legacy/src/dependency/passwords.py
"""Password hashing for unofficial (invite-joined) members.

Official members authenticate via HYPER and never set a local password.
Unofficial members set a password when they accept an invite and use it to
log back in via the dedicated unofficial login endpoint.

Uses stdlib ``hashlib.scrypt`` (memory-hard, no third-party dependency).
Hashes are stored as ``scrypt$<n>$<r>$<p>$<salt-hex>$<hash-hex>`` so params +
salt travel with the hash (verifiable & upgradable).
"""

from __future__ import annotations

import hashlib
import secrets
from typing import Tuple

# Scrypt cost parameters. N must be a power of two. These give a reasonably
# strong work factor for a low-volume internal app without being too slow.
_N = 1 << 15  # 32768
_R = 8
_P = 1
_DKEYLEN = 32
_MAXMEM = 128 * 1024 * 1024  # 128 MB headroom for scrypt's memory cost

_PREFIX = "scrypt"


def hash_password(plain: str) -> str:
    """Return a self-describing scrypt hash string for the given password."""
    salt = secrets.token_bytes(16)
    dk = hashlib.scrypt(
        plain.encode("utf-8"),
        salt=salt,
        n=_N,
        r=_R,
        p=_P,
        dklen=_DKEYLEN,
        maxmem=_MAXMEM,
    )
    return (
        f"{_PREFIX}${_N}${_R}${_P}${salt.hex()}${dk.hex()}"
    )


def _parse(stored: str) -> Tuple[int, int, int, bytes, bytes]:
    parts = stored.split("$")
    if len(parts) != 6 or parts[0] != _PREFIX:
        raise ValueError("Unrecognized password hash format")
    n, r, p = int(parts[1]), int(parts[2]), int(parts[3])
    salt = bytes.fromhex(parts[4])
    expected = bytes.fromhex(parts[5])
    return n, r, p, salt, expected


def verify_password(plain: str, stored: str) -> bool:
    """Constant-time-ish verification of a password against a stored hash."""
    try:
        n, r, p, salt, expected = _parse(stored)
    except ValueError:
        return False
    dk = hashlib.scrypt(
        plain.encode("utf-8"),
        salt=salt,
        n=n,
        r=r,
        p=p,
        dklen=len(expected),
        maxmem=_MAXMEM,
    )
    return secrets.compare_digest(dk, expected)

File: backend/src/lib/passwords.ts
import { scryptSync, randomBytes, timingSafeEqual } from "node:crypto";

/**
 * Password hashing for unofficial (invite-joined) members.
 *
 * Officials authenticate via HYPER and never set a local password. Unofficial
 * members set a password when they accept an invite and use it to log back in
 * via /auth/login-unofficial.
 *
 * Uses scrypt via `node:crypto` (available in the Workers runtime under the
 * `nodejs_compat` flag). Hashes are stored as
 *   scrypt$<N>$<r>$<p>$<salt-hex>$<hash-hex>
 * so params + salt travel with the hash (verifiable & upgradable). This is the
 * SAME format the legacy Python backend used, so existing unofficial-user
 * hashes migrate to D1 unchanged and still verify here.
 */

const PREFIX = "scrypt";
const N = 1 << 15; // 32768
const R = 8;
const P = 1;
const DKLEN = 32;
const MAXMEM = 128 * 1024 * 1024; // 128 MB headroom for scrypt's memory cost

export function hashPassword(plain: string): string {
  const salt = randomBytes(16);
  const dk = scryptSync(plain, salt, DKLEN, { N, r: R, p: P, maxmem: MAXMEM });
  return `${PREFIX}$${N}$${R}$${P}$${salt.toString("hex")}$${dk.toString("hex")}`;
}

function parse(stored: string): { n: number; r: number; p: number; salt: Buffer; expected: Buffer } {
  const parts = stored.split("$");
  if (parts.length !== 6 || parts[0] !== PREFIX) throw new Error("Unrecognized password hash format");
  const saltHex = parts[4]!;
  const hashHex = parts[5]!;
  return {
    n: Number(parts[1]),
    r: Number(parts[2]),
    p: Number(parts[3]),
    salt: Buffer.from(saltHex, "hex"),
    expected: Buffer.from(hashHex, "hex"),
  };
}

/** Verify a password against a stored scrypt hash (constant-time compare). */
export function verifyPassword(plain: string, stored: string): boolean {
  try {
    const { n, r, p, salt, expected } = parse(stored);
    const dk = scryptSync(plain, salt, expected.length, { N: n, r, p, maxmem: MAXMEM });
    return dk.length === expected.length && timingSafeEqual(dk, expected);
  } catch {
    return false;
  }
}

/** A well-formed dummy hash used to keep verify timing roughly constant when
 *  an account doesn't exist (avoids user enumeration via timing). */
export const DUMMY_HASH =
  "scrypt$32768$8$1$00000000000000000000000000000000$" +
  "0000000000000000000000000000000000000000000000000000000000000000";
